File: functions/javascript/src/setUserTier.ts
/**
 * setUserTier.ts — 管理者が任意ユーザーの tier を切り替える callable
 *
 * 用途: サポート対応・キャンペーン手動付与・検証。
 * 呼び出せるのは custom claim `admin: true` を持つユーザー、または
 * 環境変数 ADMIN_UIDS（カンマ区切り）に含まれる uid のみ。
 * tier はクライアントから直接書けない（firestore.rules で禁止）ため、
 * 切り替えは必ずこの関数を通す。
 *
 * リクエスト: { uid? , email?, tier: 'free'|'premium', duration_days?, reason?, force? }
 *   uid か email のどちらか必須。duration_days 省略時は 30 日、0 で無期限。
 * レスポンス: { uid, tier, previous_tier, expires_at }
 */
import * as functions from 'firebase-functions/v2';
import * as admin from 'firebase-admin';
import { applyTier, TierError, Tier, DEFAULT_GRANT_DAYS } from './services/tierService';

/** 環境変数 ADMIN_UIDS（カンマ区切り）の許可リスト */
function adminUids(): string[] {
  return (process.env.ADMIN_UIDS || '')
    .split(',')
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
}

function assertAdmin(auth: { uid: string; token: Record<string, unknown> } | undefined): string {
  if (!auth?.uid) {
    throw new functions.https.HttpsError('unauthenticated', '認証が必要です');
  }
  const isAdmin = auth.token?.admin === true || adminUids().includes(auth.uid);
  if (!isAdmin) {
    console.warn(`setUserTier: 権限のない呼び出し uid=${auth.uid}`);
    throw new functions.https.HttpsError('permission-denied', '管理者権限が必要です');
  }
  return auth.uid;
}

export const setUserTier = functions.https.onCall(
  {
    region: 'asia-northeast1',
    timeoutSeconds: 60,
    memory: '256MiB',
  },
  async (request) => {
    const actor = assertAdmin(request.auth);

    const {
      uid: rawUid,
      email,
      tier,
      duration_days: durationDays,
      reason,
      force,
    } = request.data as {
      uid?: string;
      email?: string;
      tier?: string;
      duration_days?: number;
      reason?: string;
      force?: boolean;
    };

    if (tier !== 'free' && tier !== 'premium') {
      throw new functions.https.HttpsError(
        'invalid-argument',
        'tier は free または premium を指定してください'
      );
    }
    if (!rawUid && !email) {
      throw new functions.https.HttpsError(
        'invalid-argument',
        'uid または email が必要です'
      );
    }
    if (
      durationDays !== undefined &&
      (typeof durationDays !== 'number' || durationDays < 0 || durationDays > 3650)
    ) {
      throw new functions.https.HttpsError(
        'invalid-argument',
        'duration_days は 0〜3650 の数値で指定してください'
      );
    }

    // email 指定はサポート対応用（問い合わせはメールアドレスで来る）
    let uid = rawUid;
    if (!uid && email) {
      try {
        uid = (await admin.auth().getUserByEmail(email)).uid;
      } catch {
        throw new functions.https.HttpsError(
          'not-found',
          `該当ユーザーが見つかりません: ${email}`
        );
      }
    }

    try {
      const result = await applyTier({
        uid: uid as string,
        tier: tier as Tier,
        durationDays: durationDays ?? DEFAULT_GRANT_DAYS,
        source: 'admin',
        actor,
        reason,
        force: force === true,
      });
      return {
        uid: result.uid,
        tier: result.tier,
        previous_tier: result.previousTier,
        expires_at: result.expiresAt,
      };
    } catch (e) {
      if (e instanceof TierError) {
        throw new functions.https.HttpsError(e.code, e.message);
      }
      console.error('setUserTier failed', e);
      throw new functions.https.HttpsError('internal', 'tier の更新に失敗しました');
    }
  }
);

File: functions/go/set_user_tier.go
package function

import (
	"context"
	"errors"
	"log"
	"os"
	"slices"
	"strings"

	"github.com/mnbst/thai-memo/functions/go/internal/callable"
	"github.com/mnbst/thai-memo/functions/go/internal/fbapp"
	"github.com/mnbst/thai-memo/functions/go/internal/tier"
)

// setUserTier は functions/javascript/src/setUserTier.ts の移植。
//
// 管理者が任意ユーザーの tier を切り替える。呼べるのは custom claim
// `admin: true` を持つユーザー、または環境変数 ADMIN_UIDS に含まれる uid のみ。
// tier はクライアントから直接書けない（firestore.rules で禁止）ため、
// 切り替えは必ずこの関数を通す。
func setUserTier(ctx context.Context, req *callable.Request) (any, error) {
	actor, err := assertAdmin(req)
	if err != nil {
		return nil, err
	}

	var in struct {
		UID          string `json:"uid"`
		Email        string `json:"email"`
		Tier         string `json:"tier"`
		DurationDays *any   `json:"duration_days"`
		Reason       string `json:"reason"`
		Force        bool   `json:"force"`
	}
	if err := req.Bind(&in); err != nil {
		return nil, err
	}

	if in.Tier != string(tier.Free) && in.Tier != string(tier.Premium) {
		return nil, callable.Errorf(callable.InvalidArgument,
			"tier は free または premium を指定してください")
	}
	if in.UID == "" && in.Email == "" {
		return nil, callable.Errorf(callable.InvalidArgument, "uid または email が必要です")
	}

	// JS 側は「未指定なら既定値、指定されたら数値かつ 0〜3650」。
	durationDays := tier.DefaultGrantDays
	if in.DurationDays != nil {
		n, ok := (*in.DurationDays).(float64)
		if !ok || n < 0 || n > 3650 {
			return nil, callable.Errorf(callable.InvalidArgument,
				"duration_days は 0〜3650 の数値で指定してください")
		}
		durationDays = int(n)
	}

	uid := in.UID
	if uid == "" {
		// email 指定はサポート対応用（問い合わせはメールアドレスで来る）
		authCli, err := fbapp.Auth(ctx)
		if err != nil {
			return nil, err
		}
		user, err := authCli.GetUserByEmail(ctx, in.Email)
		if err != nil {
			return nil, callable.Errorf(callable.NotFound,
				"該当ユーザーが見つかりません: %s", in.Email)
		}
		uid = user.UID
	}

	db, err := fbapp.Firestore(ctx)
	if err != nil {
		return nil, err
	}

	result, err := tier.ApplyTier(ctx, db, tier.Params{
		UID:          uid,
		Tier:         tier.Tier(in.Tier),
		DurationDays: durationDays,
		Source:       "admin",
		Actor:        actor,
		Reason:       in.Reason,
		Force:        in.Force,
	})
	if err != nil {
		var te *tier.Error
		if errors.As(err, &te) {
			code := callable.NotFound
			if te.Code == tier.FailedPrecondition {
				code = callable.FailedPrecondition
			}
			return nil, &callable.Error{Code: code, Message: te.Message}
		}
		log.Printf("setUserTier failed: %v", err)
		return nil, callable.Errorf(callable.Internal, "tier の更新に失敗しました")
	}

	// expires_at は無期限のとき JS では null になる。
	var expiresAt any
	if result.ExpiresAt != "" {
		expiresAt = result.ExpiresAt
	}
	return map[string]any{
		"uid":           result.UID,
		"tier":          string(result.Tier),
		"previous_tier": string(result.PreviousTier),
		"expires_at":    expiresAt,
	}, nil
}

// assertAdmin は setUserTier.ts:assertAdmin。
func assertAdmin(req *callable.Request) (string, error) {
	uid, err := req.RequireAuth()
	if err != nil {
		return "", err
	}
	if req.Auth.HasClaim("admin") || slices.Contains(adminUIDs(), uid) {
		return uid, nil
	}
	log.Printf("setUserTier: 権限のない呼び出し uid=%s", uid)
	return "", callable.Errorf(callable.PermissionDenied, "管理者権限が必要です")
}

// adminUIDs は環境変数 ADMIN_UIDS（カンマ区切り）の許可リスト。
func adminUIDs() []string {
	var out []string
	for _, s := range strings.Split(os.Getenv("ADMIN_UIDS"), ",") {
		if s = strings.TrimSpace(s); s != "" {
			out = append(out, s)
		}
	}
	return out
}
